Sources/NIOHTTP2/HTTP2PipelineHelpers.swift

```diff
@@ -164,14 +167,35 @@ extension Channel {
                                        initialLocalSettings: [HTTP2Setting] = nioDefaultSettings,
                                        position: ChannelPipeline.Position = .last,
                                        targetWindowSize: Int = 65535,
-                                       inboundStreamInitializer: ((Channel) -> EventLoopFuture<Void>)?) -> EventLoopFuture<HTTP2StreamMultiplexer> {
-        var handlers = [ChannelHandler]()
-        handlers.reserveCapacity(2)  // Update this if we need to add more handlers, to avoid unnecessary reallocation.
-        handlers.append(NIOHTTP2Handler(mode: mode, initialSettings: initialLocalSettings))
-        let multiplexer = HTTP2StreamMultiplexer(mode: mode, channel: self, targetWindowSize: targetWindowSize, inboundStreamInitializer: inboundStreamInitializer)
-        handlers.append(multiplexer)
+                                       inboundStreamInitializer: NIOChannelInitializer?) -> EventLoopFuture<HTTP2StreamMultiplexer> {
+
+        do {
+            if self.eventLoop.inEventLoop {
+                let multiplexer = try self.pipeline.syncOperations.configureHTTP2Pipeline(
+                    mode: mode,
+                    channel: self,
+                    initialLocalSettings: initialLocalSettings,
+                    position: position,
+                    targetWindowSize: targetWindowSize,
+                    inboundStreamInitializer: inboundStreamInitializer
+                )
+                return self.eventLoop.makeSucceededFuture(multiplexer)
+            } else {
+                return self.eventLoop.submit {
+                    try self.pipeline.syncOperations.configureHTTP2Pipeline(
+                        mode: mode,
+                        channel: self,
+                        initialLocalSettings: initialLocalSettings,
+                        position: position,
+                        targetWindowSize: targetWindowSize,
+                        inboundStreamInitializer: inboundStreamInitializer
+                    )
+                }
+            }
+        } catch {
+            return self.eventLoop.makeFailedFuture(error)
+        }
```

Comment:
Aim to keep do-catch blocks as small as possible. Here we should scope it to the `if self.eventLoop.inEventLoop` block.

Reply:
Ah, I missed that `self.eventLoop.submit` handles the thrown error so I thought both branches could throw and wanted to avoid duplication.